src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -127,13 +140,22 @@ private static void AllocateToken(SocketAsyncContext context, out SocketAsyncEng
         {
             lock (s_lock)
             {
-                if (s_currentEngine == null)
+                engine = s_currentEngines[s_allocateFromEngine];
+                if (engine == null)
                 {
-                    s_currentEngine = new SocketAsyncEngine();
+                    s_currentEngines[s_allocateFromEngine] = engine = new SocketAsyncEngine();
                 }
 
-                engine = s_currentEngine;
-                handle = s_currentEngine.AllocateHandle(context);
+                handle = engine.AllocateHandle(context);
+
+                if (engine.IsFull)
+                {
+                    // We'll need to create a new event port for the next handle.
+                    s_currentEngines[s_allocateFromEngine] = null;
+                }
+
+                // Round-robin to the next engine
```

Comment:
Okay, that's what I suspected. I think that this is okay for now so long as these changes don't unnecessarily penalize the scenarios that concurrently use only a small number of sockets.

I expect that in the long run a more accurate measure of the load on a socket engine will be derived from the number of socket events that occur per unit of time. Given such a metric, we may be able to periodically resort the active engines (this could be as simple as an engine moving itself to the front of an otherwise round-robin ready queue once its measured load falls below a certain threshold).